VersionControlUI/src/commands/CDiff.cpp

```diff
@@ -45,50 +40,152 @@ using namespace FilePersistence;
+	{
+		GitRepository repository{path};
 
-Interaction::CommandResult* CDiff::executeNamed(Visualization::Item*, Visualization::Item* target,
-											  const std::unique_ptr<Visualization::Cursor>&,
-											  const QString& name, const QStringList&)
+		// if there are no tokens we are unable to interpret the command
+		if (commandTokensCopy.isEmpty())
+			return false;
+
+		// check that command name starts with the characters of the first token
+		if (!name().startsWith(commandTokensCopy.takeFirst()))
+			return false;
+
+		// if there is an additional version specified, check that it is a valid commit
+		if (!commandTokensCopy.isEmpty() && !repository.isValidRevisionString(commandTokensCopy.takeFirst()))
+			return false;
+
+		// check the same if there is a second version specified
+		if (!commandTokensCopy.isEmpty() && !repository.isValidRevisionString(commandTokensCopy.takeFirst()))
+			return false;
+
+		return true;
+	}
+	else return false;
+}
+
+Interaction::CommandResult* CDiff::execute(Visualization::Item*, Visualization::Item* target,
+				const QStringList& commandTokens, const std::unique_ptr<Visualization::Cursor>& )
 {
 	auto scene = target->scene();
 	scene->clearFocus();
 	scene->clearSelection();
 	scene->setMainCursor(nullptr);
 
-	Model::TreeManager* headManager = target->node()->manager();
-	QString managerName = headManager->name();
+	QString managerName = target->node()->manager()->name();
+
 
-	// TODO add functionality to specify two versions
-	VersionControlUI::DiffManager diffManager{name, "HEAD", managerName, Model::SymbolMatcher{"Class"}};
+	// TODO restrict versions to versionA always be older?
+	// TODO diff with no versions should compare HEAD to workingDir if possible
+	// try to get the versions
+	QString versionA = commandTokens.value(1, "HEAD");
+	QString versionB = commandTokens.value(2, "HEAD");
+
+	VersionControlUI::DiffManager diffManager{versionA, versionB, managerName, Model::SymbolMatcher{"Class"}};
 	diffManager.visualize();
 
 	return new Interaction::CommandResult{};
 }
 
-QStringList CDiff::possibleNames(Visualization::Item*, Visualization::Item* target,
-											const std::unique_ptr<Visualization::Cursor>&)
+QList<Interaction::CommandSuggestion*> CDiff::suggest(Visualization::Item*, Visualization::Item* target,
+		const QString& textSoFar, const std::unique_ptr<Visualization::Cursor>&)
 {
-	Model::TreeManager* headManager = target->node()->manager();
-	QString managerName = headManager->name();
+	QStringList tokensSoFar = textSoFar.split(" ");
+
+	QList<Interaction::CommandSuggestion*> suggestions;
+	QString commandName = name();
+
+	// no suggestions for that many tokens
+	if (tokensSoFar.size() > 3)
+		return {};
+
+	// check that the command name starts with the characters of the first token
+	if (name().startsWith(tokensSoFar.takeFirst()))
+	{
+		// no additional versions specified
+		if (tokensSoFar.isEmpty())
+		{
+			// TODO implement this behavior
```

Comment:
If I understand correctly, this means that if I all I type is "diff" and then I hit enter, I get the head vs working directory.